graph/assemblygraphfast.cpp

```diff
@@ -0,0 +1,194 @@
+#include "graph/debruijnedge.h"
+#include "graph/debruijnnode.h"
+#include "graph/gfawrapper.hpp"
+#include "graph/graphicsitemedge.h"
+#include "graph/graphicsitemnode.h"
+#include "ogdf/energybased/FMMMLayout.h"
+#include "program/globals.h"
+#include "ui/myprogressdialog.h"
+
+#include <QApplication>
+#include <QList>
+#include <QRegularExpression>
+#include <QTextStream>
+
+#include <algorithm>
+#include <cmath>
+
+void AssemblyGraph::buildDeBruijnGraphFromGfaFast(const QString &fullFileName,
+                                                  bool *unsupportedCigar,
+                                                  bool *customLabels,
+                                                  bool *customColours,
+                                                  QString *bandageOptionsError) {
+    m_graphFileType = GFA;
+    m_filename = fullFileName;
+
+    QApplication::processEvents();
+    auto gfaWrapper = gfa::load(fullFileName.toStdString());
+    QApplication::processEvents();
+
+    *unsupportedCigar = false;
+    *customLabels = false;
+    *customColours = false;
+
+    /* ------------------------ NODES ------------------------ */
+    std::vector<DeBruijnNode *> nodePtrs;
+    nodePtrs.resize(gfaWrapper.verticesCount());
+    for (int64_t i = 0; i < gfaWrapper.segmentsCount(); i++) {
+        QApplication::processEvents();
+        const auto &segment = gfaWrapper.getNthSegment(i);
+        const auto &tagWrapper = segment.getTagsWrapper();
+
+        QString nodeName = segment.getName();
+        QByteArray sequence = segment.getSequence();
+
+        //We check to see if the node ended in a "+" or "-".
+        //If so, we assume that is giving the orientation and leave it.
+        //And if it doesn't end in a "+" or "-", we assume "+" and add
+        //that to the node name.
+        if (nodeName.back() != '+' && nodeName.back() != '-')
+            nodeName.push_back('+');
+
+        // Canu nodes start with "tig" which we can remove for simplicity.
+        nodeName = simplifyCanuNodeName(nodeName);
+
+        //GFA can use * to indicate that the sequence is not in the
+        //file.  In this case, try to use the LN tag for length.  If
+        //that's not available, use a length of 0.
+        //If there is a sequence, then the LN tag will be ignored.
+        int length;
+        if (sequence == "*" || sequence == "") {
+            auto lnTag = tagWrapper.getNumberTag<int>("LN");
+            if (!lnTag) {
+                throw AssemblyGraphError("expected LN tag because sequence is " + sequence);
+            }
+            length = lnTag.value();
+            sequence = "*";
+        } else
+            length = int(sequence.size());
+
+        auto dp = tagWrapper.getNumberTag<float>("DP");
+        auto kc = tagWrapper.getNumberTag<float>("KC");
+        auto rc = tagWrapper.getNumberTag<float>("RC");
+        auto fc = tagWrapper.getNumberTag<float>("FC");
+
+        double nodeDepth = 1;
```

Comment:
I'd use 0 as default value. As we cannot distinguish "good" 1x from uninitialized 1x